### PR TITLE
[Reviewer: Ellie] Stats doc changes

### PR DESCRIPTION
**Pull request by johadalin**
I've added a whole new file covering how stats are dealt with in Chronos, and an overview of the infinite table. 
Can you let me know if it generally makes sense, and also if you think this will fit as a blog post mostly as is, or if it isn't at the right level/what changes would need to be made to make it fit.

@bossmc, could you also have a look over it to see if i'm getting the table logic wrong? I followed the code as i wrote, and it is somewhat simplified, but i hope it gives a decent insight into what the table does.

### CODE REVIEW

File: doc/api.md [hunk 1]
```diff
@@ -85,6 +91,13 @@ The disadvantage of larger replication factors are:
 
 The default value for the replication factor (if unspecified) is `2`. The replication factor cannot be changed once the timer has been created.
 
+##### Statistics
+
+The `"statistics"` object is an optional parameter that may be used to provide tags holding information on what a timer represents. These tags are then used to generate stateful statistics in Chronos. The `"count"` value is also optional, defaulting to 1, but if present it must be a positive integer.
+
+Further information on querying these statistics can be found [here.](https://clearwater.readthedocs.org/en/latest/Clearwater_Stateful_Statistics/index.html)
```
Comment (eleanor-merry) [inline doc/api.md]: Should this point to the stable version?

File: doc/statistics_structures.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
+# Chronos Statistics Structures
+
+Chronos is used in Clearwater to provide stateful statistics where a deployment otherwise does not maintain state. This is done through the use of tags attached to timers. Moreover, Chronos is able to accept any tag of up to 16 A-Z characters, and can handle counts of up to the maximum value of a uint32_t.
```
Comment (eleanor-merry) [inline doc/statistics_structures.md]: Really? Where is this checked?

Comment (eleanor-merry) [inline doc/statistics_structures.md]: Explain how big this actually is

File: doc/api.md [hunk 3]
```diff
@@ -85,6 +91,13 @@ The disadvantage of larger replication factors are:
 
 The default value for the replication factor (if unspecified) is `2`. The replication factor cannot be changed once the timer has been created.
 
+##### Statistics
+
+The `"statistics"` object is an optional parameter that may be used to provide tags holding information on what a timer represents. These tags are then used to generate stateful statistics in Chronos. The `"count"` value is also optional, defaulting to 1, but if present it must be a positive integer.
```
Comment (eleanor-merry) [inline doc/api.md]: Has 'stateful statistics' been defined anywhere?

File: doc/statistics_structures.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+These values can be queried for the following time periods:
+
+* Previous five seconds
+* Current five minutes
+* Previous five minutes
+
+Chronos also allows users to query the instantaneous count of any tag.
+
+Querying statistics associated with a tag that has never been seen by Chronos will simply return `0` as the result, as through the infinite table Chronos is able to handle any tag (within specified length limits) without needing prior knowledge of its existence, enabling dynamic creation of statistics.
```
Comment (eleanor-merry) [inline doc/statistics_structures.md]: You've not mentioned what the infinite table is yet

File: doc/statistics_structures.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
+# Chronos Statistics Structures
+
+Chronos is used in Clearwater to provide stateful statistics where a deployment otherwise does not maintain state. This is done through the use of tags attached to timers. Moreover, Chronos is able to accept any tag of up to 16 A-Z characters, and can handle counts of up to the maximum value of a uint32_t.  
+This document will provide an overview of how these statistics are handled within Chronos, and how they are exposed over SNMP.
+
+## Statistics
+
+Through the functions and structures detailed below, Chronos allows the following statistics to be queried over SNMP for any tag desired:
```
Comment (eleanor-merry) [inline doc/statistics_structures.md]: This sentence reads strangely. I'd just leave out the 'Through the functions and structures detailed below, '

File: doc/statistics_structures.md (hunk 4 above)
Comment (graemerobertson) [inline doc/statistics_structures.md]: That's because nobody knows

File: doc/statistics_structures.md (hunk 2 above)
Comment (eleanor-merry) [inline doc/statistics_structures.md]: I think it's worth making it clearer up front that Chronos doesn't own the tag values - it'll just accept any tag it's given on a timer. It's therefore the clients responsibility to choose the right tag and send timer requests in such a way as to get the right statistics out

File: doc/statistics_structures.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+As detailed in the [API documentation](api.md), Chronos timer requests can be sent with a `"statistics"` object in the JSON body, specifying tags to be used in reporting statistics. These tags are parsed from the JSON, and stored as a map of `type` to `count`. The tags are then passed through to the statistics tables by the update_statistics function in the timer handler.
+
+### Tags
+
+The tags are stored as part of the timer object so that comparisons can be made on timer updates to see how the statistics have been altered. This also ensures that tag information is maintained as timers replicate across nodes.
+
+### Tables
+
+Currently we maintain two statistical views of the tags passed in as part of timers: an instantaneous count, and a time averaged set of statistics. These statistics are held in the infinite tables discussed further below. One instance of both the `infinite_scalar_table` and the `infinite_timer_count_table` are instantiated at start up. When Chronos receives a timer, we perform logic in the `update_statistics` function to determine if any statistics should be updated.
```
Comment (eleanor-merry) [inline doc/statistics_structures.md]: It's probably worth mentioning that this code is in cpp-common, not chronos

File: doc/statistics_structures.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+### Update Statistics
+
+When Chronos receives a timer to add, it first attempts to find any version of the timer within the store. If Chronos decides that it should insert the new timer into the store, either overwriting an existing timer or simply creating a new entry, the timer handler passes and tags found in both the new timer, and in any old timer, to the `update_statistics` function. N.b. If the new timer being added is to be a tombstone overwriting an existing timer, no new tags are passed in, as tombstones should not count towards our statistics.
```
Comment (eleanor-merry) [inline doc/statistics_structures.md]: I don't understand the second sentence here

Comment (eleanor-merry) [inline doc/statistics_structures.md]: is to be -> is

File: doc/statistics_structures.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+* Current five minutes
+* Previous five minutes
+
+Both of the tables simply expose increment and decrement interfaces to Chronos.
+
+### The Infinite Table
+
+Both of the above tables inherit their complex functionality from the `InfiniteBaseTable` class. The table is named as infinite because it appears to be infinite from the point of view of SNMP. i.e. SNMP can query any OID under the OID each table is registered to, and it will receive a value; if a real value is found under the table it is returned, if not a default value of 0 is returned.
+
+This is done by registering each table to a root OID (e.g. .1.2.826.0.1.1578918.999), and having the table class pass the OID of any requests under this table to internal parsing functions. The functions attempt to parse the OID of a request for the information needed to find the statistics requested by considering requests as the following format:
```
Comment (eleanor-merry) [inline doc/statistics_structures.md]: I like this section - it's a nice description of how the infinite table parses OIDs. I think it does need a up front description of *why* we need the infinite table though

File: doc/statistics_structures.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+* Previous five seconds
+* Current five minutes
+* Previous five minutes
+
+Both of the tables simply expose increment and decrement interfaces to Chronos.
+
+### The Infinite Table
+
+Both of the above tables inherit their complex functionality from the `InfiniteBaseTable` class. The table is named as infinite because it appears to be infinite from the point of view of SNMP. i.e. SNMP can query any OID under the OID each table is registered to, and it will receive a value; if a real value is found under the table it is returned, if not a default value of 0 is returned.
```
Comment (eleanor-merry) [inline doc/statistics_structures.md]: I wouldn't have the first sentence here

**Comment by eleanor-merry** [comment]
Looks good. I think that you do need some more explanation of why the stateful statistics have been implemented in the way they have been though.

**Pull request merged**